shell/platform/embedder/embedder_thread_host.cc

```diff
@@ -94,34 +116,83 @@ EmbedderThreadHost::CreateEmbedderManagedThreadHost(
     return nullptr;
   }
 
-  const auto platform_task_runner = CreateEmbedderTaskRunner(
+  // The UI and IO threads are always created by the engine and the embedder has
+  // no opportunity to specify task runners for the same.
+  // TODO(chinmaygarde): If we ever decide to expose more task runners to engine
+  // managed threads, this mask will need to be updated.
+  uint64_t engine_thread_host_mask =
+      ThreadHost::Type::UI | ThreadHost::Type::IO;
+
+  auto optional_platform_task_runner = CreateEmbedderTaskRunner(
       SAFE_ACCESS(custom_task_runners, platform_task_runner, nullptr));
+  auto optional_render_task_runner = CreateEmbedderTaskRunner(
+      SAFE_ACCESS(custom_task_runners, render_task_runner, nullptr));
+
+  if (!optional_platform_task_runner.has_value() ||
+      !optional_render_task_runner.has_value()) {
+    // User error while supplying a custom task runner. Return an invalid thread
+    // host. This will abort engine initialization. We don't want to fallback to
+    // defaults if the user wanted to specify a task runner but just messed up
+    // instead.
+    return nullptr;
+  }
 
-  // TODO(chinmaygarde): Add more here as we allow more threads to be controlled
-  // by the embedder. Create fallbacks as necessary.
+  // If the embedder has not supplied a GPU task runner, we need to create one.
+  if (!optional_render_task_runner.value()) {
+    engine_thread_host_mask |= ThreadHost::Type::GPU;
+  }
 
-  if (!platform_task_runner) {
-    return nullptr;
+  // If both the platform task runner and the GPU task runner are specified and
+  // have the same identifier, store only one.
+  if (optional_platform_task_runner.has_value() &&
```

Comment:
This should be `value()` - at this point `optional_platform_task_runner` is known to contain a value, but the value might be null.

(see the comment above - removing usage of `std::optional` here would make this clearer)

Reply:
Addressed using the pair.